### PR TITLE
Database dev

### PR DESCRIPTION
**Pull request by MaHDiaLaGaB**
Create neo4j DataBase

### CODE REVIEW

File: database.py [hunk 1]
```diff
@@ -21,27 +28,58 @@ def add_page(self, name):
...
+            "WHERE a.Name = $name"
+            "CREATE (a)-[t:Date {Time:$time}]->(b)"
+            "RETURN t;"
+        )
+        result = tx.run(query, name=page_name, post=post, time=time)
+        try:
+            return result
+        except ServiceUnavailable as exception:
+            logging.error(f"{query} raised an error: \n {exception}")
+            raise
```
Comment (MaHDiaLaGaB) [inline database.py:83]: made different methods to create relationships between page and its own posts, and it didn't work like how I wanted, so I have an idea to create relationships directly when the post node is created.

File: database.py [hunk 2]
```diff
@@ -21,27 +28,58 @@ def add_page(self, name):
...
         except Exception(exceptions.CypherSyntaxError):
             raise
 
     @staticmethod
-    def create_post_node(tx, post_text):
+    def create_post_node(tx, post_text, image):
+        query = (
+            "CREATE (b:Post {Name: 'Post', Text: $text, Photo: $image})"
+            "RETURN b"
+        )
```
Comment (MaHDiaLaGaB) [inline database.py:59]: here I will add the relationship and I will add **Time** property

**Comment by MaHDiaLaGaB** [comment]
I fixed the git issue, you can check the pull request now :)

File: scrap.py [hunk 3]
```diff
@@ -37,7 +37,10 @@ def check_connection(self) -> bool:
             raise ConnectionError(f"{self.url}: is Not reachable \nErr: {e}")
 
     # it's static function now later with database I will make it related to this class
-    def resume_scarping(self, f_posts: dict, last_post: dict) -> bool:
+
+    @staticmethod
```
Comment (Moejoe90) [inline scrap.py:41]: Why did you change this to a static method?

File: test_scrap.py [hunk 4]
```diff
@@ -22,3 +23,11 @@ def test_scraping(self):
         posts = test.scrape_full_post()
         assert posts
 
+
+
+class TestDataBase:
```
Comment (Moejoe90) [inline test_scrap.py:28]: It would make sense to move this to a separate file test_database

File: settings.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+from dotenv import load_dotenv
+from pathlib import Path
+import os
+
+load_dotenv()
+ENV_PATH = Path(__file__)/".env"
+load_dotenv(dotenv_path=ENV_PATH)
+
+NEO4J_USER = os.getenv("NEO4J_USER")
```
Comment (Moejoe90) [inline settings.py:9]: Good that you didn't post the secrets to github

File: scrap.py [hunk 6]
```diff
@@ -61,4 +64,5 @@ def scrape_full_post(self) -> dict:
                     else:
                         print("Last post already scraped")
             except exceptions.NotFound as e:
-                raise exceptions.NotFound("Post not found")
+                raise
```
Comment (Moejoe90) [inline scrap.py:67]: Why did you change the error?

File: database.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+    def __init__(self):
+        self.driver = None
+        try:
+            self.driver = GraphDatabase.driver(uri=NEO4J_URI, auth=(NEO4J_USER, NEO4J_PASSWORD))
+        except Exception(exceptions.AuthError):
+            raise
+
+    def close(self):
+        if self.driver is not None:
```
Comment (Moejoe90) [inline database.py:17]: I would throw an error here as close should never be called off the driver is not there

File: database.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            logging.error(f"{query} raised an error: \n {exception}")
+            raise
+
+    @staticmethod
+    def create_post(tx, post_text, image, time):
+        query = (
+            "CREATE (b:Post {Name: 'Post', Text: $text, Photo: $image, Time: $time}) "
+            "RETURN b"
+        )
+        result = tx.run(query, text=post_text, image=image, time=time)
```
Comment (Moejoe90) [inline database.py:95]: So here you would add a hash for all variables as I suggested

File: database.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            "RETURN t"
+        )
+        result = tx.run(query, name=page_name, time=time)
+        try:
+            return result
+        except ServiceUnavailable as exception:
+            logging.error(f"{query} raised an error: \n {exception}")
+            raise
+
+    @staticmethod
```
Comment (Moejoe90) [inline database.py:89]: I suggest putting all static methods as an abstract class

**Review by Moejoe90 — CHANGES_REQUESTED** [review]

File: scrap.py (hunk 3 above)
Comment (MaHDiaLaGaB) [reply scrap.py:41]: it's already static I just confirmed by adding the decorator

File: test_scrap.py (hunk 4 above)
Comment (MaHDiaLaGaB) [reply test_scrap.py:28]: oh ... ok nice idea

File: settings.py (hunk 5 above)
Comment (MaHDiaLaGaB) [reply settings.py:9]: I did it before and I will make sure always never happen to me again

File: scrap.py (hunk 6 above)
Comment (MaHDiaLaGaB) [reply scrap.py:67]: when an error happens it will raise twice so I did it like this to prevent duplicate

File: database.py (hunk 7 above)
Comment (MaHDiaLaGaB) [reply database.py:17]: oh you are right I didn't think of that

File: database.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+            try:
+                session.write_transaction(self.create_post, post_text, image, time)
+                print(f"the post node has been created")
+            except Exception(exceptions.SessionExpired):
+                raise
+
+    def add_relationship(self, page_name, time):
+        with self.driver.session() as session:
+            try:
+                session.write_transaction(self.create_relationship, page_name, time)
```
[the remark this answers is not in the document]
Comment (MaHDiaLaGaB) [reply database.py:42]: no, i am not with time, the time here is the bridge connecting the page and the post

File: database.py (hunk 8 above)
Comment (MaHDiaLaGaB) [reply database.py:95]: :), i don't know what is hashing but I will check it out ahahaha, it is the same as password hashing?

File: database.py (hunk 9 above)
Comment (MaHDiaLaGaB) [reply database.py:89]: ok

File: scrap.py (hunk 6 above)
Comment (Moejoe90) [reply scrap.py:67]: It's ok to have two errors it might be you need to change what the error says to reflect what happened for the context
